### DISCUSSION TITLE
Copy data model

### DISCUSSION THREAD

**Discussion by MAXPOL** [directus/directus#10589] [Archive]
Hello, i create and full config data model this name "template".
How i can copy it data model with other name ?
All settings from data model "template", but name "modelOne" and next copy again but this other name"modeTwo" and other.
Thanks.

**Comment by rijkvanzanten** [comment]
There's currently no way to clone a whole collection in data-model

**Reply by Christilut** [reply]
Is there a way to do this through the API or through a script directly on the database?

Ideally I'd like to duplicate a data model folder including all data models inside that folder. I don't mind writing a script for it, being able to duplicate a whole data model at all would save me a lot of time.

My use case is multiple websites based on the same template so I can easily deploy a new website.

**Reply by rijkvanzanten** [reply]
> My use case is multiple websites based on the same template so I can easily deploy a new website.

For this use case, I would highly recommend taking a look at [the CLI's `schema snapshot` and `schema apply` commands](https://docs.directus.io/reference/cli/#migrate-schema-to-a-different-environment). That will generate a YAML representation of your whole schema that can be used to then apply to another running instance 🙂 This will clone and spin up the whole data model though, not just a given collection.

If you're looking for a more scripted approach, you can use the [/collections](https://docs.directus.io/reference/system/collections/), [/fields](https://docs.directus.io/reference/system/fields/), and [/relations](https://docs.directus.io/reference/system/relations/) endpoints to read / write the schema (as an admin user), or go straight to the DB by cloning the tables and corresponding rows in directus_collections, _fields, and _relations respectively 👍🏻

**Reply by Christilut** [reply]
Awesome, thanks for the detailed response! :)

**Reply by Christilut** [reply]
For future users who stumble upon the same problem, here is a quick function that uses the API.
- Clones all collections & fields inside the `templateSite` parent folder
- Does nothing with relations as I did not need it

```ts
  async cloneCollection(newSite: string) {
    const templateSite = 'myParentFolder' // This is where all collections are cloned from

    const data = await this.request('GET', `/collections`)

    const collectionsToClone: string[] = data.filter(x => x.meta.group === templateSite).map(x => x.collection)

    Logger.verbose(`Cloning collections to ${newSite}`, collectionsToClone)

    // Create parent folder
    await this.request('POST', `/collections`, {
      collection: newSite,
      schema: null, // Folder
      meta: {
        collection: newSite,
        icon: 'folder'
      }
    })

    for (const oldCollectionName of collectionsToClone) {
      const data = await this.request('GET', `/collections/${oldCollectionName}`)

      Assert.truthy(data)

      const newCollectionName = data.collection.replace(`${templateSite}_`, `${newSite}_`)

      await this.request('POST', `/collections`, {
        collection: newCollectionName,
        meta: {
          ...data.meta,
          collection: newCollectionName,
          group: null
        },
        schema: {
          ...data.schema,
          name: newCollectionName
        }
      })

      await this.request('PATCH', `/collections/${newCollectionName}`, {
        meta: {
          group: newSite
        }
      })

      const fields = await this.request('GET', `/fields/${oldCollectionName}`)

      for (const field of fields) {
        // ID fields get auto created, so skip them
        if (field.field === 'id' && field.schema.default_value.includes('id_seq')) continue

        await this.request('POST', `/fields/${newCollectionName}`, {
          field: field.field,
          type: field.type,
          schema: {
            ...field.schema,
            table: null
          },
          meta: {
            ...field.meta,
            id: undefined,
            collection: null
          }
        })
      }
    }
  }
```

**Reply by mhdaouas** [reply]
Where do you call this? In which type of extension? In case of a hook, in which type of hook? Please share with us the complete code if possible. Thank you :)

**Reply by Christilut** [reply]
This was the complete code before I added relations to my directus, but after adding those it got more complicated and the code is now quite messy and custom. For relations, I basically changed the text in any reference to what it should be in the newly cloned collections. You just have to make sure to call the POSTs that create the relations in the right order.
